### DISCUSSION TITLE
Introducing flavors

### DISCUSSION THREAD

**Discussion by waghanza** [the-benchmarker/web-frameworks#3728] [General]
Hello dear team @the-benchmarker/web-frameworks,

I'll introduce `flavors` on this week end (I hope I'll have enough time).

The basic idea is to run on various supports, for example for https://github.com/phpMv/ubiquity other engine such as `swoole` or using `uWebSocket` for https://github.com/jkyberneees/0http
/cc @jcheron @jkyberneees 

However, I need an information. Which `flavor` do you (as framework author/maintainer) do you recommend ? or is mostly used in production ?

For example, as a `rails` developer, I can confirm that https://github.com/puma/puma is used in production and recommended by the community.

Regards,

**Comment by jcheron** [comment]
That's a great initiative! :heart_eyes:
- `Swoole` seems to me essential, (used alone or combined with others such as Laravel, symfony...).
- `Workerman` a lot of associated Frameworks and the advantage of having nothing to install ~~(just a composer require ...)~~
- [Roadrunner](https://github.com/spiral/roadrunner) (associated with Spiral, Ubiquity, but also with others)
- [Ngx_php](https://github.com/rryqszq4/ngx_php7) extension module associated with nginx, more recent but very promising .

~~When taking the Ubiquity case, Swoole is the only one that requires a special extension. For the others (Roadrunner, Workerman or ngx_php), just add a composer dependency.~~

**Reply by leocavalcante** [reply]
Actually, is not that trivial.
- *Workerman:* for a fair production-grade comparison you have to install `ext-event`.
- *RoadRunner:* requires the `rr` binary itself, but yeah, a `./vendor/bin/rr get-binary` will do it for you.
- *ngx_php:* requires recompiling Nginx with `ngx_php7` module that depends on the `php-src` itself.

They are not just a Composer dependency. Workman can work like that, but is not the recommended production way, its for development purposes, like running PHP's built-in webserver (`php -S`).

**And I'd like to add to the list [`ReactPHP`](https://reactphp.org/) and [`Amp`](https://amphp.org/).**

Ps.: they looks like you are just requiring a Composer dependency as well, but for a production-grade environment it is highly recommended that you run their event-loops based on **extensions** that uses high-performant syscalls like `epoll` instead of `select`.

Like Amp documentation says:
> It’s perfectly fine to have one of the **extensions** in production while relying on the NativeDriver locally for development.

- [Amp](https://amphp.org/amp/event-loop/#implementations)
- [ReactPHP](https://reactphp.org/event-loop/#loop-implementations)

**Comment by jcheron** [comment]
Yes, I was wrong to downplay the work. Thank you for rectifying @leocavalcante .
In production, Workerman requires the compiled `ext-event`.
And ngx_php can't do without the nginx compilation.

**Comment by waghanza** [comment]
Yeah, I plan that (except for `ngx_php7`), but more important I need your insight @jcheron @leocavalcante

What do you recommend in production (not about performances, but more usage, safety ...).

**Reply by leocavalcante** [reply]
I'd say Swoole, ReactPHP and Amp, personally I'm not seeing much of Workerman, RoadRunner and ngx_php.
